### ISSUE TITLE
Alphanumeric produce different cite keys in text and bibliography

### ISSUE THREAD

**Issue by tkw1536** [typst/hayagriva#255]
### Description

I am using the `alphanumeric` citation style. When citing two entries from a bib file that would produce the same key (e.g. same author and same year), they render with an `a` and `b` suffix where they are cited. However, in the bibliography, they appear without the same suffix. 
Without manual tweaking, with `#show` rules and the like I was expecting the same label to be used in both places. 

As a minimal example, consider the following bib file:

```bib
@article{first,
  title={First Thing},
  year={2024},
  author={Jane Doe}
}

@article{second,
  title={Second Thing},
  year={2024},
  author={Jane Doe}
}
```

The `alphanumeric` cite styles assigns both articles the key `Doe24`. 

Using the following main file:

```typst
#set cite(style: "alphanumeric")

I cite the first thing #cite(<first>) and the second thing #cite(<second>). 

#bibliography("example.bib")
```

The pdf renders as:

<img width="435" alt="Screenshot 2024-12-08 at 12 54 30" src="https://github.com/user-attachments/assets/c99ff515-5780-4fa3-819c-6c65fa31a30b" alt="I cite the first thing [Doe24a] and [Doe24b]. Bibliography. [Doe24] J. Doe, First Thing, 2024a. [Doe24] J. Doe, Second Thing, 2024a">

Notice how the keys listed in the main body are [Doe24a] and [Doe24b] whereas in the bibliography section they are both shown as [Doe24]. There also appears to be an extra an and b at the end of the entries. 

I did find two related issues, but neither of them seem to cover exactly this.
- typst/typst#3986, but I'm not using `bibliography.full`
- typst/typst#2707, but that's closed and this is still a problem

### Reproduction URL

https://typst.app/project/rY9ipQQ8EYDQ3hjCEXMy22

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by PgBiel** [comment]
Seems to be a regression from https://github.com/typst/hayagriva/pull/234

**Comment by Marcono1234** [comment]
https://docs.citationstyles.org/en/stable/specification.html#disambiguation ("`disambiguate-add-year-suffix` [Method (4)]") says regarding this:
> By default, the year-suffix is appended the first year rendered through `cs:date` in the cite and in the bibliographic entry, but its location can be controlled by explicitly rendering the “year-suffix” variable using `cs:text`. If “year-suffix” is rendered through `cs:text` in the scope of `cs:citation`, it is suppressed for `cs:bibliography`, unless it is also rendered through `cs:text` in the scope of `cs:bibliography`, and vice versa.

I might be misunderstanding it, but to me it seems the short form date in the citation label is not recognized as date. Though on the other hand I am not sure if the CSL specification requires an explicit use of `cs:date` or if the implicit use through the citation label is enough.

Maybe Hayagriva's [`alphanumeric.csl`](https://github.com/typst/hayagriva/blob/main/styles/alphanumeric.csl) also wouldn't need the explicit `<text variable="year-suffix" />` then; in fact according to the specification quoted above the implicit use of `year-suffix` for the cite should suppress implicit usage of the suffix in the bibliography (which is currently not happening for Hayagriva either?).



So maybe a proper workaround for now is using a custom CSL file, e.g. something like this (based on IEEE):
<details>
<summary>CSL file snippet (click to expand)</summary>
(Might not follow best practices)

```xml
...
  <!-- Citation label with disambiguation suffix, see related https://github.com/typst/hayagriva/issues/255 -->
  <macro name="citation-label-with-suffix">
    <group>
      <text variable="citation-label"/>
      <text variable="year-suffix"/>
    </group>
  </macro>
  <!-- Citation -->
  <!-- Based on https://github.com/typst/hayagriva/pull/307, not yet available in latest Typst release-->
  <citation collapse="citation-number" after-collapse-delimiter="; "
  disambiguate-add-year-suffix="true">
    <sort>
      <key variable="author"/>
      <key variable="issued"/>
    </sort>
    <layout prefix="[" suffix="]" delimiter=", ">
      <group delimiter=", ">
        <text macro="citation-label-with-suffix"/>
        <text variable="locator"/>
      </group>
    </layout>
  </citation>
  <!-- Bibliography -->
  <bibliography entry-spacing="0">
    <sort>
      <key macro="citation-label-with-suffix"/>
    </sort>
    <layout>
      <!-- Citation Key -->
      <text macro="citation-label-with-suffix" prefix="[" suffix="]" display="left-margin"/>
      ... <!-- remaining bib entry data -->
```

</details>

**Comment by filchristou** [comment]
Are there any good news for typst 0.14.2 ?

**Comment by tkw1536** [comment]
I don't think so, I have just re-tried the minimal example I provided above and it still produces the same broken result.

**Comment by filchristou** [comment]
fyi thanks to [this comment](https://forum.typst.app/t/alphanumeric-style-how-to-get-disambiguation-letters-e-g-ass96a-and-ass96b-in-bibliography-labels/3871/14?u=filchristou), I found out that pergamon supports this. However I am not sure about hayagriva; I went back to bibtex.
